### ISSUE TITLE
Could NameNode HA be supported?

### ISSUE THREAD

**Issue by tomsheep** [mtth/hdfs#18]
That will be great if the `url` parameter for `Client` accepted multiple host:port pairs.

**Comment by mtth** [comment]
Thanks for the suggestion. This seems like a non-trivial change however, especially since I'm not familiar with this HA implementation and can't easily test against it.

**Comment by abridgett** [comment]
It's not great (no retry, failover) but as a simple approach I just check status to find which one is up (if the namenode is in standby then it'll reject the request):

```
    for url in HDFS_URLS:
        hadoop = hdfs.InsecureClient(url)
        try:
            hadoop.status('/')
            return url
        except:
            continue
    logging.critical("Unable to contact any HDFS namenode")
    exit(1)
```

**Comment by marty90** [comment]
i would be interested in this feature too

**Comment by sansagara** [comment]
I'm actually needing this. Important specially in case of Namenode failures, and Active/Standby switching. In this case, all the scripts are going to fail unless you change the default NN on `hdfscli.cfg`.  My main NN broke this morning, so the script hitting it failed with error:

`  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/util/connection.py", line 88, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused`

I could catch this exception, and then fire a change on the default namenode (client) on the OS with os.utils (something like a rename or maybe directly changing the relevant string). Going to work on this, and let you know. Another different thing is if Both NN are up, but your configuration is hitting the Standby one. That brings an HdfsError.

BTW, what i have right now is something like: 

> [global]
> default.alias = prod1
> autoload.modules = hdfs.ext.kerberos
> 
> [prod1.alias]
> url = http://nn1.company.com:50000/
> client = KerberosClient
> 
> [prod2.alias]
> url = http://nn2.company.com:50000/
> client = KerberosClient

So i manually change between `prod1` and `prod2` in default.alias as needed. However, i would appreciate any recommendation on how to do this switch directly on Python. (A script or a config).

**Comment by sansagara** [comment]
Another thing on this: For those using Cloudera Manager for their clusters, maybe looking at this is helpful:
http://www.cloudera.com/documentation/enterprise/5-3-x/topics/cdh_hag_hdfs_ha_admin.html#concept_md3_l2r_bp__section_wyz_bkj_wn

**Comment by sansagara** [comment]
Ok, so another update: I ended up using the HttpFS service directly instead of hitting the NN directly. It turns out that HttpFS will receive the request and route it to the active NN. So, no need to worry about that yourself. Reason i was not using that on the first place, is i did not know how to get that info. It's actually on Cloudera Manager. You go to:

- Hosts tab > All Hosts
- Filter by `Services > HUE`
- Expand the different roles. You are looking for Role `HDFS HttpFS`. Click on it.
- Configuration (tab) 
- Ports and Addresses (Toolbar). Write the url/port there.

**Comment by KevinFTD** [comment]
Using HttpFS is another way to avoid this problem.

**Comment by ndimiduk** [comment]
Please not another daemon/gateway process.

**Comment by dmorel** [comment]
for the record, AFAICT trying both hosts in a HA pair and failing over to the active one (so it succeeds 50% of the time on the first hit) is what the java hdfs client lib does, which produces logs that are misleading to the casual observer.

**Comment by mtth** [comment]
Released in `2.1.0`, credit to @adaitche (see https://github.com/mtth/hdfs/pull/80).
